### PR TITLE
[BugFix] Fix race condition between preloading partial update state and publish

### PR DESCRIPTION
**Pull request by TszKitLo40**
Why I'm doing:
For partial update, preloading partial update will read pk index while publishing will update pk index. if there is no lock is used, memory problem will occur.

What I'm doing:
Use lock in `get_rowids_from_pkindex` and publishing.
Fixes #issue

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/src/storage/lake/lake_primary_index.h [hunk 1]
```diff
@@ -46,13 +46,17 @@ class LakePrimaryIndex : public PrimaryIndex {
     int64_t data_version() const { return _data_version; }
     void update_data_version(int64_t version) { _data_version = version; }
 
+    std::mutex* mutex() { return &_mutex; }
```
Comment (kevincai) [inline be/src/storage/lake/lake_primary_index.h:49]: can't be a worse design to expose a mutex directly to outside.

File: be/src/storage/lake/txn_log_applier.cpp [hunk 2]
```diff
@@ -49,6 +49,8 @@ class PrimaryKeyTxnLogApplier : public TxnLogApplier {
     ~PrimaryKeyTxnLogApplier() override {
         // must release primary index before `handle_failure`, otherwise `handle_failure` will fail
         _tablet.update_mgr()->release_primary_index_cache(_index_entry);
+        auto& index = _index_entry->value();
+        index.unlock();
```
Comment (luohaha) [inline be/src/storage/lake/txn_log_applier.cpp:53]: If `prepare_primary_index` fail, and we don't get lock yet, we don't need to unlock, right?

File: be/src/storage/lake/update_manager.cpp [hunk 3]
```diff
@@ -81,6 +81,7 @@ StatusOr<IndexEntry*> UpdateManager::prepare_primary_index(const TabletMetadata&
         LOG(ERROR) << msg;
         return Status::InternalError(msg);
     }
+    index.lock();
```
Comment (luohaha) [inline be/src/storage/lake/update_manager.cpp:84]: We need to lock before load.

File: be/src/storage/lake/update_manager.cpp [hunk 4]
```diff
@@ -340,6 +342,22 @@ Status UpdateManager::get_rowids_from_pkindex(Tablet* tablet, int64_t base_versi
                                               const std::vector<ColumnUniquePtr>& upserts,
                                               std::vector<std::vector<uint64_t>*>* rss_rowids) {
     Status st;
+    st.update(_handle_index_op(tablet, base_version, [&](LakePrimaryIndex& index) {
+        // get rss_rowids for each segment of rowset
+        index.fetch_guard();
```
Comment (luohaha) [inline be/src/storage/lake/update_manager.cpp:347]: Can this work?

File: be/src/storage/lake/update_manager.cpp [hunk 5]
```diff
@@ -355,6 +373,22 @@ Status UpdateManager::get_rowids_from_pkindex(Tablet* tablet, int64_t base_versi
...
+        uint32_t num_segments = upserts.size();
+        for (size_t i = 0; i < num_segments; i++) {
+            auto& pks = *upserts[i];
+            st.update(index.get(pks, &((*rss_rowids)[i])));
+        }
+    }));
+    return st;
+}
+
+Status UpdateManager::get_rowids_from_pkindex_unlock(Tablet* tablet, int64_t base_version,
```
Comment (luohaha) [inline be/src/storage/lake/update_manager.cpp:388]: No need to add `xx_lock` and `xx_unlock`.

**Review by luohaha — DISMISSED** [review]

File: be/src/storage/lake/update_manager.cpp [hunk 6]
```diff
@@ -329,16 +331,20 @@ Status UpdateManager::_handle_index_op(Tablet* tablet, int64_t base_version,
     if (!index.is_load(base_version)) {
         return Status::Uninitialized(fmt::format("Primary index not load yet, tablet_id: {}", tablet->id()));
     }
+    std::unique_ptr<std::lock_guard<std::mutex>> guard = nullptr;
+    if (need_lock) {
+        guard = index.fetch_guard();
```
Comment (luohaha) [inline be/src/storage/lake/update_manager.cpp:336]: fetch guard with timeout.

**Review by luohaha — APPROVED** [review]

**Review by wyb — APPROVED** [review]

**Pull request merged**